src/Hamiltonians/abstract.jl

```diff
@@ -289,3 +289,24 @@ function Base.Matrix(h::AbstractHamiltonian, args...; kwargs...)
     return Matrix(BasisSetRep(h, args...; kwargs...))
 end
 Base.Matrix(bsr::BasisSetRep) = Matrix(bsr.sm)
+
+# TODO: implement guiding vector importance sampling
+"""
+    TransformOperator{T,K<:AbstractHamiltonian,O<:Union{AbstractHamiltonian,Nothing}}
+
+Type for creating operators defined by a transformation `transform`. Can create an operator
+that is a transformation of an input `op`, or an operator that is only defined by `transform`.
+
+Not exported; transformations should define properties of these operators before use.
+
+Currently supported transformations:
+
+    * [`GutzwillerSampling`](@ref).
+"""
+struct TransformOperator{T,K<:AbstractHamiltonian,O<:Union{AbstractHamiltonian,Nothing}} <: AbstractHamiltonian{T}
+    transform::K
+    op::O
+end
```

Comment:
I very much like having docstrings even for types or functions that are not exported just as a way of documenting what they do. This one reads a bit cryptic, though, at the moment.

My suggestion here would be to rename the type to `UndoSimilarityTransform`. I don't think the function `similarity_transform` is needed (see my comment there). Instead I would document here how to use the constructor for this type (which would implement what `similarity_transform` does right now). 

Instead of `nothing::Nothing`, we might use `I::UniformScaling{Bool}` (available trough the `LinearAlgebra` standard library) for the operator in the case we just want `f^{-2}`. `I` is a super cool way to represent a universal identity matrix, which is exactly what we need here.

The specialisation of the functionality of the type to `GutzwillerSampling` happens naturally by dispatching on the type of `K`, as it is already implemented with the new methods in `GutzwillerSampling.jl`. In order to implement other importance sampling schemes, such as `GuidingVectorSampling`, we just need to write the respective methods. E.g.
```
diagonal_element(s::UndoSimilarityTransform{<:Any,<:GuidingVectorSampling,<:AbstractHamiltonian}, add)
```

Reply:
Yes the name can change, but I don't think the type should be specific to similarity transformations. We may want to have some other kind of transformation in the future.

As for the function `similarity_transform` I was trying the suggestion by @mtsch to avoid creating an instance of a transformed operator that is not yet supported. I originally just had the constructor without a wrapper function, but perhaps I misunderstood what he meant?

Reply:
Actually, I wanted to suggest `UndoImportanceSampling` for the name. This is the problem at hand, and the type (and its name) is general enough to capture both Gutzwiller and guiding vector sampling. 

The main point is that we are not really transforming the observable `op` in any meaningful way. We are annotating `op` such that in a three-way dot product with right-hand-side eigenvectors, we know how to undo the importance sampling transformation that was applied to them.

I'm not sure whether we can undo more general similarity transforms in this way (e.g. the transcorrelated transformation is tricky to undo). I'm thinking that we'd need a different mechanism, and that could come with a dedicated type.

What other kinds of transformation do you have in mind?

Reply:
Regarding @mtsch's comment, we need to discuss. As long as the type is unexported, and the docstring is clear, I'm not too worried about error messages when people try to do something non-sensical. We could create an abstract type for `ImportanceSamplingTransform` and restrict the constructor this way, but I'm not sure it is needed.

Reply:
Ah, sorry for causing confusion. I agree we should discuss this.